utils/env_preferences/src/windows.rs

```diff
@@ -0,0 +1,37 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use windows::{Globalization, System::UserProfile::GlobalizationPreferences};
+
+use crate::RetrievalError;
+pub fn get_locales() -> Result<Vec<String>, RetrievalError> {
+    let mut locale_vec_str: Vec<String> = Vec::new();
+    let locale = GlobalizationPreferences::Languages();
+
+    match locale {
+        Ok(languages) => {
+            for i in 0..languages.Size().unwrap() {
+                let hstring = languages.GetAt(i).unwrap();
+                let string: String = hstring.to_string_lossy();
+                locale_vec_str.push(string);
+            }
+        }
+        Err(e) => return Err(RetrievalError::Other(e.to_string())),
```

Comment:
This seems to be the only place where you use `Error::Other`. Generally `Other` is lossy - when as a customer of your API I see `Other` the way you defined it I can't reason much about it - I can only display it to the user.

But you don't need to make it lossy. You can add a variant called `Windows` to your error which embeds https://docs.rs/windows-result/0.2.0/windows_result/struct.Error.html which is an Error returned by all Windows.rs crates.

```
pub enum RetrievalError {
    /// Unable to retrieve the locale
    NullLocale,
    ...
    Windows(windows_result::Error),
    Other(String),
}
```

Next, you can implement a `From`:

```
impl From<windows_result::Error> for RetrievalError {
    fn from(input: windows_result::Error) -> Self {
        Self::Windows(input)
    }
}
```

Once you have this, in this `get_locales` you can use `?` which will auto convert and propagate the error:

```
let locale = GlobalizationPreferences::Languages()?;

for i in 0.. languages.Size()? {
    let hstring = languages.GetAt(i)?;
    let string = hstring.to_string_lossy();
    locale_vec_str.push(string);
}
```

And once you're here, the iterator combinators are very powerful in Rust (similar to Python) and optimize really well. You start with a [`IVectorView`](https://microsoft.github.io/windows-docs-rs/doc/windows/System/UserProfile/struct.GlobalizationPreferences.html#method.Languages).

It implements a trait called `IntoIterator`, great! So we can iterate over the elements without having to rely on Size and then GetAt pair:

```
let input_locales = GlobalizationPreferences::Languages()?;
let result_locales: Vec<String> = languages
    .into_iter()
    .map(|hstring| { hstring.to_string_lossy() })
    .collect();

Ok(result_locales)
```

Reply:
Great take @zbraniecki